### PR TITLE
Fix high severity issues CS-4007

### PR DESCRIPTION
**Pull request by FadhlanR**
Tickets:  CS-4007
Issues: #24 , #25 , #26 , #27 , #28 , #29 , #30 , #31 , #32 , #33 , #34 , #35 , #36 , #37

### CODE REVIEW

**Review by jurgenwerk — APPROVED** [review]
Adding @alex-cardstack and @habdelra to review this. I'm not competent enough yet to approve some of these fixes, for example the re-entrancy fix. Other stuff looks good to me.

**Review by habdelra — APPROVED** [review]
I think this looks great! @alex-cardstack, should approve as well.

File: contracts/libraries/SafeERC721.sol [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    /**
+    * @dev Same as ERC721.transfer(address,uint256) but with extra consistency checks.
+    * @param _token address of the token contract
+    * @param _to address of the receiver
+    * @param _value amount of tokens to send
+    */
+    function safeTransfer(address _token, address _to, uint256 _value) internal {
+        ERC721(_token).transfer(_to, _value);
+        assembly {
```
Comment (alex-cardstack) [inline contracts/libraries/SafeERC721.sol:22]: So, the assembly code in this new file - is it copypasted from the ERC20 library?

I guess it kind of makes sense what it is doing - it's copying the return data from the previous call if there is any and reverting if it's zero.

I guess this is so it handles contracts that have a `transfer` function that don't return _anything_?

It looks like [there's a much nicer way to do this in more modern solidity](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/utils/SafeERC20.sol#L105-L115) but I think this seems to make sense on first glance

**Review by alex-cardstack — APPROVED** [review]

File: contracts/libraries/SafeERC721.sol (hunk 1 above)
Comment (FadhlanR) [reply contracts/libraries/SafeERC721.sol:22]: Yes, it's copypasted from the ERC20 library, for now, the objective is to check the transfer result. Agree for next improvement we can use the openzeppelin contract.

**Pull request merged**